packages/service-library/tests/fastapi/test_request_decorators.py

```diff
@@ -0,0 +1,248 @@
+    async def _correct_signature(_request: Request, x: int):
+        ...
+
+    with pytest.raises(ValueError) as exc_info:
+
+        @app.get("/wrong")
+        @cancellable_request
+        async def _wrong_signature():
+            ...
+
+    assert "_wrong_signature" in str(exc_info.value)
+
+
+async def test_with_non_cancellable_entrypoint(
+    test_client: TestClient, inspect_app_tasks: Callable[[], list[Task]]
+):
+    assert not inspect_app_tasks()
+
+    # NOT cancellable entrypoint
+    r = await test_client.get("/")
+    assert r.status_code == status.HTTP_200_OK
+
+    assert not inspect_app_tasks()
+
+
+async def test_cancellable_handler_with_successful_request(
+    test_client: TestClient, inspect_app_tasks: Callable[[], list[Task]]
+):
+    # cancellable entrypoint
+    r = await test_client.get("/sleep/0")
+    assert r.status_code == status.HTTP_200_OK
+
+    assert all(
+        t.done() for t in inspect_app_tasks()
+    ), "All tasks (if any) should be done"
+
+
+#
+# NOTE:
+# - Cannot find a way to disconnect a client using the TestClient fixture  :-(
+# - Any ideas welcome!
+#
+
+
+@pytest.mark.skip(reason="DEV: cannot emulate disconnect")
+async def test_it(
+    app: FastAPI,
+    event_loop: asyncio.AbstractEventLoop,
+    inspect_app_tasks: Callable[[], list[Task]],
+):
+    #
+    # Extension of TestClient to send a disconnect request
+    #
+    # It does send a disconnect request but the monitor task
+    # interprets it as "separate" and does not cancel the
+    # initial
+    #
+    class TestClientExt(TestClient):
+        def disconnect(
+            self,
+            path: str,
+            *,
+            method: str = "GET",
+            scheme: str = "http",
+        ):
+            input_queue: asyncio.Queue[dict] = asyncio.Queue()
+            output_queue: asyncio.Queue[dict] = asyncio.Queue()
+            scope = {
+                "type": "http",
+                "http_version": "1.1",
+                "asgi": {"version": "3.0"},
+                "method": method,
+                "scheme": scheme,
+                "path": path,
+                "query_string": b"",
+                "root_path": "",
+                "headers": flatten_headers({"Connection": "close"}),
+            }
+            running_task = create_monitored_task(
+                self.application(scope, input_queue.get, output_queue.put),
+                output_queue.put_nowait,
+            )
+
+            send = input_queue.put_nowait
+            send({"type": "http.disconnect"})
+
+    # tests ---
+
+    # on_start event not called
+    with pytest.raises(AttributeError) as exc_info:
+        assert app.state.started
+    assert "started" in f"{exc_info.value}"
+
+    # Emulating disconnect
+    async with TestClientExt(app) as client:
+        assert app.state.started
+        assert not app.state.stopped
+
+        client_task = event_loop.create_task(
+            client.get("/sleep/100", headers={"Connection": "close"})
```

Comment:
I guess you tried it, but what about passing a timeout of 0.000000001 seconds in the client directly? this should just disconnect super fast. 
If you have a request entrypoint that sleeps for 1 seconds it should be good?

Reply:
yes. Pair reviewed.

Reply:
Found interesting example https://github.dev/RedRoserade/fastapi-disconnect-example
But there they use manual testing (i.e. spawns server and client separately by hand)  to verify. We might have to do something like that with processes to achieve that at the expense of loosing control on the test/debugging side.